libraries/lib-concurrency/ParallelReduce.h

```diff
@@ -0,0 +1,307 @@
+            assert(prevTask->begin() == oldFirst);
+            assert(prevTask->end() == first);
+            prevTask = &tasks.back();
+            assert(prevTask->begin() == first);
+            assert(prevTask->end() == end);
+         }
+      }
+   }
+   n_threads = tasks.size() + 1;
+
+   // We will need futures as communication channels among tasks
+   using Packaged = std::packaged_task<void(Task&, size_t)>;
+   using Future = std::future<void>;
+   using Channel = std::pair<Packaged, Future>;
+   std::vector<Channel> channels;
+   channels.reserve(tasks.size());
+
+   // The runner captures a reference to the array of not-yet-constructed
+   // channels
+   Progress<Increment> progress;
+   const auto runner =
+   [n_threads, &tasks, &channels, &progress](Task &task, size_t index){
+      // Typically the task will iterate over its sub-range
+      std::exception_ptr exc;
+      bool aborting = false;
+      try {
+         task.run(progress);
+      }
+      catch(const detail::AbortException &) {
+         // Some other thread aborted; don't skip the merging below, which must
+         // propagate the original exception.  Don't propagate AbortException.
+         aborting = true;
+      }
+      catch(...) {
+         aborting = true;
+         // Cause other threads to go through the previous catch when they test
+         // progress
+         progress.abandon();
+         exc = std::current_exception();
+      }
+      // Merge results of some other tasks with higher index, in a "butterfly"
+      // (or "hypercube") pattern
+      for (size_t step = 1; !(index & step); step <<= 1) {
+         auto neighborIndex = index + step;
+         if (neighborIndex >= n_threads)
+            break;
+         --neighborIndex;
+         // Synchronize on the completion of the other task before merging
+         // (or, propagate an exception)
+         try {
+            channels[neighborIndex].second.get();
+         }
+         catch(...) {
+            aborting = true;
+            // Remember only one exception
+            exc = std::current_exception();
+         }
+         if (!aborting)
+            try {
+               task.merge(std::move(tasks[neighborIndex]), progress);
+            }
+            catch(...) {
+               aborting = true;
+               // Remember only one exception
+               exc = std::current_exception();
+            }
+      }
+      if (exc)
+         rethrow_exception(exc);
+   };
+
+   // Make the channels, but don't start threads before all futures exist!
+   for (auto &task : tasks) {
+      auto pt = Packaged{ std::cref(runner) };
+      auto future = pt.get_future();
+      channels.emplace_back(move(pt), move(future));
+   }
+
+   // Now it's safe to run the new tasks concurrently, for all indices more than
+   // 0
+   size_t index = 0;
+   for (auto &task : tasks) {
+      std::thread{ ref(channels[index].first), std::ref(task), index + 1 }
```

Comment:
Contrary to the previous place, I see very little reason to split this into multiple lines instead of using ++index right in place.

Reply:
Wrong -- `index` is used twice in arguments of a constructor and order of evaluation is not specified.  I must increment only in another statement.